subprojects/core/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java

```diff
@@ -326,43 +323,65 @@ private LockInfo readInformationRegion(CountdownTimer timer) throws IOException,
-                    } else {
-                        LOGGER.debug("The file lock is held by a different Gradle process. I was unable to read on which port the owner listens for lock access requests.");
+        private java.nio.channels.FileLock lockStateRegion(final LockMode lockMode, final CountdownTimer timer) throws IOException, InterruptedException {
+            return retryUntil(timer, new IOQuery<java.nio.channels.FileLock>() {
+                @Override
+                public java.nio.channels.FileLock run() throws IOException, InterruptedException {
+                    java.nio.channels.FileLock fileLock = lockFileAccess.tryLockState(lockMode == LockMode.Shared);
+                    if (fileLock != null) {
+                        return fileLock;
+                    }
+                    if (port != -1) { //we don't like the assumption about the port very much
+                        LockInfo lockInfo = readInformationRegion(timer);
+                        if (lockInfo.port != -1) {
+                            LOGGER.debug("The file lock is held by a different Gradle process (pid: {}, operation: {}). Will attempt to ping owner at port {}", lockInfo.pid, lockInfo.operation, lockInfo.port);
+                            fileLockContentionHandler.pingOwner(lockInfo.port, lockInfo.lockId, displayName);
+                        } else {
+                            LOGGER.debug("The file lock is held by a different Gradle process. I was unable to read on which port the owner listens for lock access requests.");
+                        }
                     }
+                    return null;
                 }
-                //TODO SF we should inform on the progress/status bar that we're waiting
-                Thread.sleep(backoff(++i));
-            } while (!timer.hasExpired());
-            return null;
+            });
         }
 
-        private long backoff(int i) {
-            return random.nextInt(Math.min(i, EXPONENTIAL_BACKOFF_CAP_FACTOR)) * SLOT_TIME;
+        private java.nio.channels.FileLock lockInformationRegion(final LockMode lockMode, CountdownTimer timer) throws IOException, InterruptedException {
+            return retryUntil(timer, new IOQuery<java.nio.channels.FileLock>() {
+                @Override
+                public java.nio.channels.FileLock run() throws IOException {
+                    return lockFileAccess.tryLockInfo(lockMode == LockMode.Shared);
+                }
+            });
         }
 
-        private java.nio.channels.FileLock lockInformationRegion(LockMode lockMode, CountdownTimer timer) throws IOException, InterruptedException {
-            int i = 0;
-            do {
-                java.nio.channels.FileLock fileLock = lockFileAccess.tryLockInfo(lockMode == LockMode.Shared);
-                if (fileLock != null) {
-                    return fileLock;
+        private <T> T retryUntil(CountdownTimer timer, IOQuery<T> query) throws IOException, InterruptedException {
```

Comment:
I would move this method to `ExponentialBackoff` too, it would avoid the `forIteration` timeout to leak and keep the logic where it belongs to.